### ISSUE TITLE
Inspect potential race in the network package

### ISSUE THREAD

**Issue by Liamsi** [dedis/cothority#646]
Error/data race which happens rarely in tests of the network package
<details><summary>

travis output from build</summary>

```
3 : (               network.(*Router).connect: 180) - tcp://[::]:2007 Connecting to  local://127.1.2.3:2890
3 : (               network.(*Router).connect: 184) - Could not connect to  local://127.1.2.3:2890 TCPHost local://127.1.2.3:2890 can't handle this type of connection: local
3 : (               network.(*Router).connect: 180) - tcp://[::]:2007 Connecting to  tcp://127.0.0.1:2008
3 : (               network.(*Router).connect: 188) - tcp://[::]:2007 Connected to  tcp://127.0.0.1:2008
4 : (    network.(*Router).registerConnection: 278) - tcp://[::]:2007 Registers tcp://127.0.0.1:2008
4 : (                  network.(*Router).Send: 153) - tcp://[::]:2007 sends to tcp://127.0.0.1:2008 msg: &{I:12}
3 : (            network.(*Router).handleConn: 219) - tcp://[::]:2007 Handling new connection to  tcp://127.0.0.1:2008
4 : ( network.(*Router).receiveServerIdentity: 373) - tcp://[::]:2008 Identity received from  tcp://127.0.0.1:2007
4 : (    network.(*Router).registerConnection: 278) - tcp://[::]:2008 Registers tcp://127.0.0.1:2007
3 : (            network.(*Router).handleConn: 219) - tcp://[::]:2008 Handling new connection to  tcp://127.0.0.1:2007
4 : (            network.(*Router).handleConn: 232) - tcp://127.0.0.1:2008 got error (Connection Closed) while receiving message
3 : (            network.(*Router).handleConn: 237) - tcp://[::]:2008 handleConn with closed connection: stop (dst= tcp://127.0.0.1:2007 )
4 : (                  network.(*Router).Send: 153) - tcp://[::]:2007 sends to tcp://127.0.0.1:2008 msg: &{I:12}
2 : (                  network.(*Router).Send: 158) - tcp://[::]:2007 Couldn't send to tcp://127.0.0.1:2008 : Connection Closed trying again
3 : (               network.(*Router).connect: 180) - tcp://[::]:2007 Connecting to  tcp://127.0.0.1:2008
4 : (            network.(*Router).handleConn: 232) - tcp://127.0.0.1:2007 got error (EOF) while receiving message
3 : (            network.(*Router).handleConn: 237) - tcp://[::]:2007 handleConn with closed connection: stop (dst= tcp://127.0.0.1:2008 )
3 : (               network.(*Router).connect: 184) - Could not connect to  tcp://127.0.0.1:2008 Could not connect to tcp://127.0.0.1:2008: dial tcp 127.0.0.1:2008: getsockopt: connection refused
3 : (               network.(*Router).connect: 180) - local://127.0.0.1:2007 Connecting to  local://127.1.2.3:2890
3 : (               network.(*Router).connect: 184) - Could not connect to  local://127.1.2.3:2890 Could not connect 6c6f63616c3a2f2f3132372e302e302e313a323030372063616e277420636f6e6e65637420746f206c6f63616c3a2f2f3132372e312e322e333a323839303a2069742773206e6f74206c697374656e696e67
3 : (               network.(*Router).connect: 180) - local://127.0.0.1:2007 Connecting to  local://127.0.0.1:2008
3 : (               network.(*Router).connect: 188) - local://127.0.0.1:2007 Connected to  local://127.0.0.1:2008
4 : (    network.(*Router).registerConnection: 278) - local://127.0.0.1:2007 Registers local://127.0.0.1:2008
4 : (                  network.(*Router).Send: 153) - local://127.0.0.1:2007 sends to local://127.0.0.1:2008 msg: &{I:12}
3 : (            network.(*Router).handleConn: 219) - local://127.0.0.1:2007 Handling new connection to  local://127.0.0.1:2008
4 : ( network.(*Router).receiveServerIdentity: 373) - local://127.0.0.1:2008 Identity received from  local://127.0.0.1:2007
4 : (    network.(*Router).registerConnection: 278) - local://127.0.0.1:2008 Registers local://127.0.0.1:2007
3 : (            network.(*Router).handleConn: 219) - local://127.0.0.1:2008 Handling new connection to  local://127.0.0.1:2007
4 : (            network.(*Router).handleConn: 232) - local://127.0.0.1:2007 got error (Connection Closed) while receiving message
3 : (            network.(*Router).handleConn: 237) - local://127.0.0.1:2007 handleConn with closed connection: stop (dst= local://127.0.0.1:2008 )
4 : (                  network.(*Router).Send: 153) - local://127.0.0.1:2007 sends to local://127.0.0.1:2008 msg: &{I:12}
2 : (                  network.(*Router).Send: 158) - local://127.0.0.1:2007 Couldn't send to local://127.0.0.1:2008 : Connection Closed trying again
3 : (               network.(*Router).connect: 180) - local://127.0.0.1:2007 Connecting to  local://127.0.0.1:2008
3 : (               network.(*Router).connect: 184) - Could not connect to  local://127.0.0.1:2008 Could not connect 6c6f63616c3a2f2f3132372e302e302e313a323030372063616e277420636f6e6e65637420746f206c6f63616c3a2f2f3132372e302e302e313a323030383a2069742773206e6f74206c697374656e696e67
3 : (               network.(*Router).connect: 180) - tcp://[::]:2009 Connecting to  tcp://127.0.0.1:2010
3 : (               network.(*Router).connect: 188) - tcp://[::]:2009 Connected to  tcp://127.0.0.1:2010
4 : (    network.(*Router).registerConnection: 278) - tcp://[::]:2009 Registers tcp://127.0.0.1:2010
4 : (                  network.(*Router).Send: 153) - tcp://[::]:2009 sends to tcp://127.0.0.1:2010 msg: &{I:3}
3 : (            network.(*Router).handleConn: 219) - tcp://[::]:2009 Handling new connection to  tcp://127.0.0.1:2010
4 : ( network.(*Router).receiveServerIdentity: 373) - tcp://[::]:2010 Identity received from  tcp://127.0.0.1:2009
4 : (    network.(*Router).registerConnection: 278) - tcp://[::]:2010 Registers tcp://127.0.0.1:2009
3 : (            network.(*Router).handleConn: 219) - tcp://[::]:2010 Handling new connection to  tcp://127.0.0.1:2009
4 : (                  network.(*Router).Send: 153) - tcp://[::]:2010 sends to tcp://127.0.0.1:2009 msg: &{I:3}
3 : (               network.(*Router).connect: 180) - tcp://[::]:2001 Connecting to  tcp://127.0.0.1:2000
3 : (               network.(*Router).connect: 180) - tcp://[::]:2000 Connecting to  tcp://127.0.0.1:2001
3 : (               network.(*Router).connect: 188) - tcp://[::]:2000 Connected to  tcp://127.0.0.1:2001
4 : (    network.(*Router).registerConnection: 278) - tcp://[::]:2000 Registers tcp://127.0.0.1:2001
4 : (                  network.(*Router).Send: 153) - tcp://[::]:2000 sends to tcp://127.0.0.1:2001 msg: &{I:3}
3 : (            network.(*Router).handleConn: 219) - tcp://[::]:2000 Handling new connection to  tcp://127.0.0.1:2001
3 : (               network.(*Router).connect: 188) - tcp://[::]:2001 Connected to  tcp://127.0.0.1:2000
4 : (    network.(*Router).registerConnection: 278) - tcp://[::]:2001 Registers tcp://127.0.0.1:2000
4 : (                  network.(*Router).Send: 153) - tcp://[::]:2001 sends to tcp://127.0.0.1:2000 msg: &{I:3}
3 : (            network.(*Router).handleConn: 219) - tcp://[::]:2001 Handling new connection to  tcp://127.0.0.1:2000
4 : ( network.(*Router).receiveServerIdentity: 373) - tcp://[::]:2001 Identity received from  tcp://127.0.0.1:2000
4 : (    network.(*Router).registerConnection: 278) - tcp://[::]:2001 Registers tcp://127.0.0.1:2000
3 : (            network.(*Router).handleConn: 219) - tcp://[::]:2001 Handling new connection to  tcp://127.0.0.1:2000
4 : ( network.(*Router).receiveServerIdentity: 373) - tcp://[::]:2000 Identity received from  tcp://127.0.0.1:2001
4 : (    network.(*Router).registerConnection: 278) - tcp://[::]:2000 Registers tcp://127.0.0.1:2001
3 : (            network.(*Router).handleConn: 219) - tcp://[::]:2000 Handling new connection to  tcp://127.0.0.1:2001
3 : (               network.(*Router).connect: 180) - local://127.0.0.1:2001 Connecting to  local://127.0.0.1:2000
3 : (               network.(*Router).connect: 188) - local://127.0.0.1:2001 Connected to  local://127.0.0.1:2000
4 : (    network.(*Router).registerConnection: 278) - local://127.0.0.1:2001 Registers local://127.0.0.1:2000
4 : (                  network.(*Router).Send: 153) - local://127.0.0.1:2001 sends to local://127.0.0.1:2000 msg: &{I:3}
4 : ( network.(*Router).receiveServerIdentity: 373) - local://127.0.0.1:2000 Identity received from  local://127.0.0.1:2001
4 : (    network.(*Router).registerConnection: 278) - local://127.0.0.1:2000 Registers local://127.0.0.1:2001
3 : (            network.(*Router).handleConn: 219) - local://127.0.0.1:2001 Handling new connection to  local://127.0.0.1:2000
3 : (               network.(*Router).connect: 180) - local://127.0.0.1:2000 Connecting to  local://127.0.0.1:2001
3 : (            network.(*Router).handleConn: 219) - local://127.0.0.1:2000 Handling new connection to  local://127.0.0.1:2001
3 : (               network.(*Router).connect: 188) - local://127.0.0.1:2000 Connected to  local://127.0.0.1:2001
4 : (    network.(*Router).registerConnection: 278) - local://127.0.0.1:2000 Registers local://127.0.0.1:2001
4 : ( network.(*Router).receiveServerIdentity: 373) - local://127.0.0.1:2001 Identity received from  local://127.0.0.1:2000
4 : (    network.(*Router).registerConnection: 278) - local://127.0.0.1:2001 Registers local://127.0.0.1:2000
3 : (            network.(*Router).handleConn: 219) - local://127.0.0.1:2000 Handling new connection to  local://127.0.0.1:2001
4 : (                  network.(*Router).Send: 153) - local://127.0.0.1:2000 sends to local://127.0.0.1:2001 msg: &{I:3}
3 : (            network.(*Router).handleConn: 219) - local://127.0.0.1:2001 Handling new connection to  local://127.0.0.1:2000
4 : (            network.(*Router).handleConn: 232) - local://127.0.0.1:2001 got error (Connection Closed) while receiving message
3 : (            network.(*Router).handleConn: 237) - local://127.0.0.1:2001 handleConn with closed connection: stop (dst= local://127.0.0.1:2000 )
4 : (            network.(*Router).handleConn: 232) - local://127.0.0.1:2001 got error (Connection Closed) while receiving message
3 : (            network.(*Router).handleConn: 237) - local://127.0.0.1:2001 handleConn with closed connection: stop (dst= local://127.0.0.1:2000 )
3 : (               network.(*Router).connect: 180) - tcp://[::]:2011 Connecting to  tcp://127.0.0.1:2012
3 : (               network.(*Router).connect: 188) - tcp://[::]:2011 Connected to  tcp://127.0.0.1:2012
4 : ( network.(*Router).receiveServerIdentity: 373) - tcp://[::]:2012 Identity received from  tcp://127.0.0.1:2011
4 : (    network.(*Router).registerConnection: 278) - tcp://[::]:2012 Registers tcp://127.0.0.1:2011
3 : (            network.(*Router).handleConn: 219) - tcp://[::]:2012 Handling new connection to  tcp://127.0.0.1:2011
4 : (    network.(*Router).registerConnection: 278) - tcp://[::]:2011 Registers tcp://127.0.0.1:2012
4 : (                  network.(*Router).Send: 153) - tcp://[::]:2011 sends to tcp://127.0.0.1:2012 msg: &{I:5}
3 : (            network.(*Router).handleConn: 219) - tcp://[::]:2011 Handling new connection to  tcp://127.0.0.1:2012
2 : (         network.testRouterSendMsgDuplex: 296) - Received msg h1 -> h2 {5}
4 : (                  network.(*Router).Send: 153) - tcp://[::]:2012 sends to tcp://127.0.0.1:2011 msg: &{I:5}
2 : (         network.testRouterSendMsgDuplex: 303) - Received msg h2 -> h1 {5}
3 : (               network.(*Router).connect: 180) - local://127.0.0.1:2011 Connecting to  local://127.0.0.1:2012
3 : (               network.(*Router).connect: 188) - local://127.0.0.1:2011 Connected to  local://127.0.0.1:2012
4 : (    network.(*Router).registerConnection: 278) - local://127.0.0.1:2011 Registers local://127.0.0.1:2012
4 : ( network.(*Router).receiveServerIdentity: 373) - local://127.0.0.1:2012 Identity received from  local://127.0.0.1:2011
4 : (    network.(*Router).registerConnection: 278) - local://127.0.0.1:2012 Registers local://127.0.0.1:2011
3 : (            network.(*Router).handleConn: 219) - local://127.0.0.1:2012 Handling new connection to  local://127.0.0.1:2011
4 : (                  network.(*Router).Send: 153) - local://127.0.0.1:2011 sends to local://127.0.0.1:2012 msg: &{I:5}
3 : (            network.(*Router).handleConn: 219) - local://127.0.0.1:2011 Handling new connection to  local://127.0.0.1:2012
2 : (         network.testRouterSendMsgDuplex: 296) - Received msg h1 -> h2 {5}
4 : (                  network.(*Router).Send: 153) - local://127.0.0.1:2012 sends to local://127.0.0.1:2011 msg: &{I:5}
2 : (         network.testRouterSendMsgDuplex: 303) - Received msg h2 -> h1 {5}
4 : (            network.(*Router).handleConn: 232) - local://127.0.0.1:2012 got error (Connection Closed) while receiving message
3 : (            network.(*Router).handleConn: 237) - local://127.0.0.1:2012 handleConn with closed connection: stop (dst= local://127.0.0.1:2011 )
3 : (               network.(*Router).connect: 180) - tcp://[::]:8787 Connecting to  tcp://127.0.0.1:7878
4 : ( network.(*Router).receiveServerIdentity: 373) - tcp://[::]:7878 Identity received from  tcp://127.0.0.1:8787
4 : (    network.(*Router).registerConnection: 278) - tcp://[::]:7878 Registers tcp://127.0.0.1:8787
3 : (            network.(*Router).handleConn: 219) - tcp://[::]:7878 Handling new connection to  tcp://127.0.0.1:8787
3 : (               network.(*Router).connect: 188) - tcp://[::]:8787 Connected to  tcp://127.0.0.1:7878
4 : (    network.(*Router).registerConnection: 278) - tcp://[::]:8787 Registers tcp://127.0.0.1:7878
4 : (                  network.(*Router).Send: 153) - tcp://[::]:8787 sends to tcp://127.0.0.1:7878 msg: &{I:12}
3 : (            network.(*Router).handleConn: 219) - tcp://[::]:8787 Handling new connection to  tcp://127.0.0.1:7878
4 : (              network.TestRouterExchange: 347) - Closing connections
4 : (            network.(*Router).handleConn: 232) - tcp://127.0.0.1:7878 got error (EOF) while receiving message
3 : (            network.(*Router).handleConn: 237) - tcp://[::]:7878 handleConn with closed connection: stop (dst= tcp://127.0.0.1:8787 )
4 : ( network.(*Router).receiveServerIdentity: 373) - tcp://[::]:7878 Identity received from  tcp://127.0.0.1:8787
4 : (    network.(*Router).registerConnection: 278) - tcp://[::]:7878 Registers tcp://127.0.0.1:8787
E : (           network.(*Router).Start.func1:  71) - receive server identity failed: Error while receiving ServerIdentity during negotiation EOF
==================
WARNING: DATA RACE
Write at 0x00c420270478 by goroutine 54:
  internal/race.Write()
      /home/travis/.gimme/versions/go1.7.linux.amd64/src/internal/race/race.go:41 +0x38
  sync.(*WaitGroup).Wait()
      /home/travis/.gimme/versions/go1.7.linux.amd64/src/sync/waitgroup.go:129 +0x151
  github.com/dedis/cothority/network.(*Router).Stop()
      github.com/dedis/cothority/network/_test/_obj_test/router.go:119 +0x377
  github.com/dedis/cothority/network.TestRouterExchange()
      /home/travis/gopath/src/github.com/dedis/cothority/network/router_test.go:351 +0x60e
  github.com/dedis/cothority/network.TestRouterExchange()
      /home/travis/gopath/src/github.com/dedis/cothority/network/router_test.go:342 +0x4fd
  github.com/dedis/cothority/network.(*Router).Send()
      github.com/dedis/cothority/network/_test/_obj_test/router.go:145 +0x828
  github.com/dedis/cothority/network.TestRouterExchange()
      /home/travis/gopath/src/github.com/dedis/cothority/network/router_test.go:330 +0x42c
  github.com/dedis/cothority/network.TestRouterExchange()
      /home/travis/gopath/src/github.com/dedis/cothority/network/router_test.go:325 +0x3a1
  testing.tRunner()
      /home/travis/.gimme/versions/go1.7.linux.amd64/src/testing/testing.go:610 +0xc9
Previous read at 0x00c420270478 by goroutine 57:
  internal/race.Read()
      /home/travis/.gimme/versions/go1.7.linux.amd64/src/internal/race/race.go:37 +0x38
  sync.(*WaitGroup).Add()
      /home/travis/.gimme/versions/go1.7.linux.amd64/src/sync/waitgroup.go:71 +0x292
  github.com/dedis/cothority/network.(*Router).launchHandleRoutine()
      github.com/dedis/cothority/network/_test/_obj_test/router.go:292 +0x63
  github.com/dedis/cothority/network.(*Router).Start.func1()
      github.com/dedis/cothority/network/_test/_obj_test/router.go:82 +0x3fc
Goroutine 54 (running) created at:
  testing.(*T).Run()
      /home/travis/.gimme/versions/go1.7.linux.amd64/src/testing/testing.go:646 +0x52f
  testing.RunTests.func1()
      /home/travis/.gimme/versions/go1.7.linux.amd64/src/testing/testing.go:793 +0xb9
  testing.tRunner()
      /home/travis/.gimme/versions/go1.7.linux.amd64/src/testing/testing.go:610 +0xc9
  testing.RunTests()
      /home/travis/.gimme/versions/go1.7.linux.amd64/src/testing/testing.go:799 +0x4ba
  testing.(*M).Run()
      /home/travis/.gimme/versions/go1.7.linux.amd64/src/testing/testing.go:743 +0x12f
  github.com/dedis/cothority/log.MainTest.func1()
      /home/travis/gopath/src/github.com/dedis/cothority/log/lvl.go:386 +0x38
Goroutine 57 (finished) created at:
  github.com/dedis/cothority/network.(*TCPListener).Listen.func1()
      github.com/dedis/cothority/network/_test/_obj_test/tcp.go:360 +0x85
  github.com/dedis/cothority/network.(*TCPListener).listen()
      github.com/dedis/cothority/network/_test/_obj_test/tcp.go:401 +0x38c
  github.com/dedis/cothority/network.(*TCPListener).Listen()
      github.com/dedis/cothority/network/_test/_obj_test/tcp.go:363 +0xba
  github.com/dedis/cothority/network.(*TCPHost).Listen()
      <autogenerated>:93 +0x72
  github.com/dedis/cothority/network.(*Router).Start()
      github.com/dedis/cothority/network/_test/_obj_test/router.go:83 +0xfe
  github.com/dedis/cothority/network.TestRouterExchange.func1()
      /home/travis/gopath/src/github.com/dedis/cothority/network/router_test.go:316 +0x65
==================
3 : (            network.(*Router).handleConn: 219) - tcp://[::]:7878 Handling new connection to  tcp://127.0.0.1:8787
E : (network.(*ServerIdentityToml).ServerIdentity: 142) - Error while reading public key: EOF
1 : (         network.TestTCPConnReceiveRaw.func2: 155) - Will write slice %d/%d... 0 6
1 : (         network.TestTCPConnReceiveRaw.func2: 159) -  OK
1 : (         network.TestTCPConnReceiveRaw.func2: 155) - Will write slice %d/%d... 1 6
1 : (         network.TestTCPConnReceiveRaw.func2: 159) -  OK
1 : (         network.TestTCPConnReceiveRaw.func2: 155) - Will write slice %d/%d... 2 6
1 : (         network.TestTCPConnReceiveRaw.func2: 159) -  OK
1 : (         network.TestTCPConnReceiveRaw.func2: 155) - Will write slice %d/%d... 3 6
1 : (         network.TestTCPConnReceiveRaw.func2: 159) -  OK
1 : (         network.TestTCPConnReceiveRaw.func2: 155) - Will write slice %d/%d... 4 6
1 : (         network.TestTCPConnReceiveRaw.func2: 159) -  OK
1 : (         network.TestTCPConnReceiveRaw.func2: 166) - Will write last piece...
1 : (         network.TestTCPConnReceiveRaw.func2: 170) -  OK
3 : (                    network.TestTCPHostClose: 368) - Finished first connection, starting 2nd
3 : (                    network.TestTCPHostClose: 378) - Closing h3
PASS
coverage: 89.5% of statements
Found 1 data race(s)
FAIL    github.com/dedis/cothority/network  4.507s
```

</details>

**Comment by nikkolasg** [comment]
Another Travis build timeout in `TestRouterLotsOfConnLocal`.

**Comment by nikkolasg** [comment]
This is being fixed in #673
